usr/sbin/pkcsslotd/slotmgr.c

```diff
@@ -329,182 +327,213 @@ static int create_pid_file(pid_t pid)
-                 keyword_token_to_str(tok));
-		return -1;
-	}
-    memset(slotinfo, padding, size);
-	memcpy(slotinfo, val, strlen(val));
-	return 0;
+    return 0;
 }
 
-static int do_vers(struct parse_data *d,
-                   CK_VERSION *slotinfo, int kw, const char *val)
+static int config_parse(const char *config_file)
 {
-	char *dot;
-
-	if (!val || !*val) {
-		snprintf(d->errbuf, sizeof(d->errbuf), "%s has no value\n",
-                 keyword_token_to_str(kw));
-		return -1 ;
-	}
-
-    dot = strchr(val, '.');
-    slotinfo->major = strtol(val, NULL, 10);
-    slotinfo->minor = dot ? strtol(dot + 1, NULL, 10) : 0;
-	return 0;
-}
+    FILE *file;
+    struct ConfigBaseNode *c, *config = NULL;
+    struct ConfigIdxStructNode *slot;
+    int i, ret = 0;
+
+    file = fopen(config_file, "r");
+    if (file == NULL) {
+        ErrLog("Error opening config file '%s': %s\n", config_file,
+               strerror(errno));
+        return -1;
+    }
 
-static int slotmgr_key_str(void *private, int tok, const char *val)
-{
-    struct parse_data *d = (struct parse_data *)private;
-    CK_VERSION vers;
-    int res = 0;
-
-    switch (tok) {
-    case KW_STDLL:
-        if (do_str(d, (char *)&d->sinfo_struct.dll_location,
-                   sizeof(d->sinfo_struct.dll_location), tok, val, 0))
-            res = 1;
-        else
-            d->sinfo_struct.present = TRUE;
-        break;
-    case KW_SLOTDESC:
-        if (do_str(d, (char *)d->sinfo_struct.pk_slot.slotDescription,
-                   sizeof(d->sinfo_struct.pk_slot.slotDescription), tok, val, ' '))
-            res = 1;
-        break;
-    case KW_MANUFID:
-        if (do_str(d, (char *)d->sinfo_struct.pk_slot.manufacturerID,
-                   sizeof(d->sinfo_struct.pk_slot.manufacturerID), tok, val, ' '))
-            res = 1;
-        break;
-    case KW_CONFNAME:
-        if (do_str(d, (char *)d->sinfo_struct.confname,
-                   sizeof(d->sinfo_struct.confname), tok, val, 0))
-            res = 1;
-        break;
-    case KW_TOKNAME:
-        if (do_str(d, (char *)d->sinfo_struct.tokname,
-                   sizeof(d->sinfo_struct.tokname), tok, val, 0))
-            res = 1;
-        break;
-    case KW_HWVERSION:
-        if (do_vers(d, &d->sinfo_struct.pk_slot.hardwareVersion, tok, val))
-            res = 1;
-        break;
-    case KW_FWVERSION:
-        if (do_vers(d, &d->sinfo_struct.pk_slot.firmwareVersion, tok, val))
-            res = 1;
-        break;
-    case KW_TOKVERSION:
-        if (do_vers(d, &vers, tok, val))
-            res = 1;
-        else
-            d->sinfo_struct.version = vers.major << 16 | vers.minor;
-        break;
-    default:
-        snprintf(d->errbuf, sizeof(d->errbuf),
-                 "Unknown string-valued keyword detected: \"%s\"",
-                 keyword_token_to_str(tok));
-        res = 1;
-        break;
+    ret = parse_configlib_file(file, &config, config_parse_error, 0);
```

Comment:
Not sure if we want to do that, but seeing that the configuration contains paths to libraries that we dlopen, should we check file permissions on the config file?  If it is world-writable, we should probably not use it.

Reply:
well, we didn't check that when using the old parser either....
Nevertheless, it might certainly be a good ting to add such a check.  Could it have any (unwanted) negative impact on existing installations?

Reply:
Depending on the modes we enforce, yes.  Assume we enforce that only root has write permissions but the distribution does not install it that way.

Reply:
I have added a 4th commit to check if the file is world writable:
pkcsslotd: 
- pkcsslotd: Check for opencryptoki.conf file permissions

It does not check for specific owner or group, as we don't know how this file is installed by distributors.